vertx-service-discovery-grpc/src/main/java/io/vertx/servicediscovery/grpc/impl/GrpcServiceImpl.java

```diff
@@ -0,0 +1,116 @@
+ */
+public class GrpcServiceImpl<T> implements GrpcService {
+
+  @Override
+  public String name() {
+    return TYPE;
+  }
+
+  @Override
+  public ServiceReference get(Vertx vertx, ServiceDiscovery discovery, Record record,
+                              JsonObject conf) {
+    Objects.requireNonNull(vertx);
+    Objects.requireNonNull(record);
+    Objects.requireNonNull(discovery);
+    return new GrpcServiceImpl.GrpcServiceReference(vertx, discovery, record, null);
+  }
+
+  /**
+   * Implementation of {@link ServiceReference} for event bus service proxies.
+   */
+  public class GrpcServiceReference extends AbstractServiceReference<T> {
+
+    private final String stub;
+    private final VertxChannelBuilder builder;
+
+    /**
+     * Creates a new instance of {@link GrpcServiceReference}.
+     *
+     * @param vertx      the vert.x instance, must not be {@code null}
+     * @param discovery  the service discovery instance, must not be {@code null}
+     * @param record     the service record, must not be {@code null}
+     * @param customizer the method used to customize the GRPC channel
+     */
+    GrpcServiceReference(Vertx vertx, ServiceDiscovery discovery, Record record,
+                         Handler<VertxChannelBuilder> customizer) {
+      super(vertx, discovery, record);
+      this.stub = record.getMetadata().getString("stub.class");
+      Objects.requireNonNull(stub);
+
+      builder = VertxChannelBuilder
+        .forAddress(vertx,
+          record().getLocation().getString("host"),
+          record().getLocation().getInteger("port")
+        );
+
+      if (customizer != null) {
+        customizer.handle(builder);
+      }
+    }
+
+    /**
+     * Methods used to customize the channel.
+     *
+     * @param handler the customizer method
+     * @return the current reference
+     */
+    public GrpcServiceReference customizeChannel(Handler<VertxChannelBuilder> handler) {
+      handler.handle(builder);
+      return this;
+    }
+
+    /**
+     * Build the stub object and return it. If already built, it returns the cached one.
+     *
+     * @return the stub instance
+     */
+    @Override
+    public synchronized T retrieve() {
+      Class<T> itf = ClassLoaderUtils.load(stub, this.getClass().getClassLoader());
+      if (itf == null) {
+        throw new IllegalStateException("Cannot load stub class " + stub);
+      } else {
+        // Create the channel
+        ManagedChannel channel = builder.build();
+
+        Class[] args = new Class[]{io.grpc.Channel.class};
+        try {
+          Constructor<T> constructor;
+          try {
+            constructor = itf.getConstructor(args);
+          } catch (NoSuchMethodException ignore) {
+            // Try to find the private version of the constructor.
+            constructor = itf.getDeclaredConstructor(args);
+            if (!constructor.isAccessible()) {
+              constructor.setAccessible(true);
```

Comment:
With respect to Java9 shouldn't we just patch the compiler instead? to avoid this visibility hack